File: API/3497_More_Orders_After_BreakEven/CS/MoreOrdersAfterBreakEvenStrategy.cs
namespace StockSharp.Samples.Strategies;

using System;
using Ecng.Common;
using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.Messages;

/// <summary>
/// More Orders After BreakEven strategy: TEMA crossover.
/// Buys when price crosses above TEMA, sells when price crosses below TEMA.
/// </summary>
public class MoreOrdersAfterBreakEvenStrategy : Strategy
{
	private readonly StrategyParam<DataType> _candleType;
	private readonly StrategyParam<int> _period;

	private decimal _prevClose;
	private decimal _prevTema;
	private bool _hasPrev;

	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }
	public int Period { get => _period.Value; set => _period.Value = value; }

	public MoreOrdersAfterBreakEvenStrategy()
	{
		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
			.SetDisplay("Candle Type", "Candle timeframe", "General");
		_period = Param(nameof(Period), 14)
			.SetGreaterThanZero()
			.SetDisplay("Period", "TEMA period", "Indicators");
	}

	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);
		_hasPrev = false;
		var tema = new TripleExponentialMovingAverage { Length = Period };
		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(tema, ProcessCandle).Start();
	}

	private void ProcessCandle(ICandleMessage candle, decimal temaValue)
	{
		if (candle.State != CandleStates.Finished) return;

		if (_hasPrev)
		{
			if (_prevClose <= _prevTema && candle.ClosePrice > temaValue && Position <= 0)
				BuyMarket();
			else if (_prevClose >= _prevTema && candle.ClosePrice < temaValue && Position >= 0)
				SellMarket();
		}

		_prevClose = candle.ClosePrice;
		_prevTema = temaValue;
		_hasPrev = true;
	}
}

File: API/3401-3500/3497_More_Orders_After_BreakEven/PY/more_orders_after_break_even_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage
from StockSharp.Algo.Strategies import Strategy

class more_orders_after_break_even_strategy(Strategy):
    def __init__(self):
        super(more_orders_after_break_even_strategy, self).__init__()

        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(60)))
        self._period = self.Param("Period", 50)

        self._prev_close = 0.0
        self._prev_tema = 0.0
        self._has_prev = False

    @property
    def CandleType(self):
        return self._candle_type.Value

    @CandleType.setter
    def CandleType(self, value):
        self._candle_type.Value = value

    @property
    def Period(self):
        return self._period.Value

    @Period.setter
    def Period(self, value):
        self._period.Value = value

    def OnReseted(self):
        super(more_orders_after_break_even_strategy, self).OnReseted()
        self._prev_close = 0.0
        self._prev_tema = 0.0
        self._has_prev = False

    def OnStarted2(self, time):
        super(more_orders_after_break_even_strategy, self).OnStarted2(time)
        self._prev_close = 0.0
        self._prev_tema = 0.0
        self._has_prev = False

        tema = ExponentialMovingAverage()
        tema.Length = self.Period

        subscription = self.SubscribeCandles(self.CandleType)
        subscription.Bind(tema, self._process_candle).Start()

    def _process_candle(self, candle, tema_value):
        if candle.State != CandleStates.Finished:
            return

        close = float(candle.ClosePrice)
        tema_val = float(tema_value)

        if self._has_prev:
            if self._prev_close <= self._prev_tema and close > tema_val and self.Position <= 0:
                self.BuyMarket()
            elif self._prev_close >= self._prev_tema and close < tema_val and self.Position >= 0:
                self.SellMarket()

        self._prev_close = close
        self._prev_tema = tema_val
        self._has_prev = True

    def CreateClone(self):
        return more_orders_after_break_even_strategy()
